static/compiler-revision-picker.ts

```diff
@@ -0,0 +1,216 @@
+        }
+
+        const item = this.tomSelect.items[0];
+        if (typeof item === 'undefined') {
+            return;
+        }
+
+        const option = this.tomSelect.options[item];
+        const nextOffset = option.offset + distance;
+
+        if (nextOffset < 0) {
+            // We know this offset is invalid.
+            return;
+        }
+
+        // TODO: Check whether we have this offset locally.
+        const res = await searchCompilerLibrary({
+            compilerId: this.lastCompilerId,
+            offset: option.offset + distance,
+            // Request two revisions. We can detect whether we've exhaused the
+            // list of revisions by checking whether we only get 1 in response.
+            limit: 2,
+            query: '',
+        });
+        const data = await res.json();
+        if (data.results.length === 0) {
+            // Offset is past the end of the list.
+            return;
+        }
+        if (nextOffset === 0) {
+            // TODO: Disable prev button
+        }
+        if (data.results.length === 1) {
+            // TODO: Disable next button
+        }
+
+        const nextRevision = data.results[0];
+
+        // Insert the option into the picker and set it as active.
+        if (Object.prototype.hasOwnProperty.call(this.tomSelect.options, 'nextRevision')) {
+            this.tomSelect.addOptions([
+                {
+                    revision: nextRevision,
+                    offset: nextOffset,
+                },
+            ]);
+        }
+        this.tomSelect.removeItem(item);
+        this.tomSelect.addItem(nextRevision);
+    }
+
+    initialize(compilerId: string) {
+        this.lastCompilerId = compilerId;
+
+        // eslint-disable-next-line @typescript-eslint/no-this-alias
+        const self = this;
+        this.tomSelect = new TomSelect($(this.domNode).find('select')[0], {
+            valueField: 'revision',
+            labelField: 'revision',
+            searchField: ['revision'],
+            placeholder: '🔍 Select revision...',
+            optgroupField: '$groups',
+            lockOptgroupOrder: true,
+            items: compilerId ? [compilerId] : [],
+            dropdownParent: 'body',
+            closeAfterSelect: true,
+            plugins: ['dropdown_input', 'virtual_scroll'],
+            maxOptions: 1000,
+            firstUrl: function (query) {
+                return '0';
+            },
+            load: async function (query, callback) {
+                const offset = parseInt((this as any).getUrl(query));
+                const res = await searchCompilerLibrary({
+                    compilerId: self.lastCompilerId,
+                    query,
+                    offset,
+                });
+                const data = await res.json();
+                const items = data.results.map((revision, i) => ({
+                    revision,
+                    // Track the offset of the revision in the backend result
```

Comment:
nice. (I think maybe your comment settings are wrapping too early, seems like `set.` could fit on the same line as the prior? supernitpick!)

Reply:
The line below was wrapped at 80 and I hadn't looked at prettier config. 120 chars is a very respectable line length.